argmin/src/solver/quasinewton/lbfgs.rs

```diff
@@ -138,28 +164,167 @@ where
         self.tol_cost = tol_cost;
         Ok(self)
     }
+
+    /// Activates L1-regularization with coefficient `l1_coeff`.
+    ///
+    /// Parameter `l1_coeff` must be `> 0.0`.
+    ///
+    /// # Example
+    ///
+    /// ```
+    /// # use argmin::solver::quasinewton::LBFGS;
+    /// # use argmin::core::Error;
+    /// # fn main() -> Result<(), Error> {
+    /// # let linesearch = ();
+    /// let lbfgs: LBFGS<_, Vec<f64>, Vec<f64>, f64> = LBFGS::new(linesearch, 3).with_l1_regularization(1.0)?;
+    /// # Ok(())
+    /// # }
+    pub fn with_l1_regularization(mut self, l1_coeff: F) -> Result<Self, Error> {
+        if l1_coeff <= float!(0.0) {
+            return Err(argmin_error!(
+                InvalidParameter,
+                "`L-BFGS`: coefficient of L1-regularization must be > 0."
+            ));
+        }
+        self.l1_coeff = Some(l1_coeff);
+        Ok(self)
+    }
+}
+
+/// Wrapper problem for supporting constrained line search.
+struct LineSearchProblem<O, P, G, F> {
+    problem: O,
+    xi: Option<P>,
+    l1_coeff: Option<F>,
+    phantom: PhantomData<G>,
+}
+
+impl<O, P, G, F> LineSearchProblem<O, P, G, F>
+where
+    P: ArgminSub<F, P>,
+    F: ArgminFloat,
+{
+    fn new(problem: O) -> Self {
+        Self {
+            problem,
+            xi: None,
+            l1_coeff: None,
+            phantom: PhantomData,
+        }
+    }
+
+    fn with_l1_constraint(&mut self, l1_coeff: F, param: &P, pseudo_gradient: &G)
+    where
+        P: ArgminZeroLike
+            + ArgminMinMax
+            + ArgminSignum
+            + ArgminAdd<P, P>
+            + ArgminAdd<F, P>
+            + ArgminMul<P, P>
+            + ArgminSub<F, P>
+            + ArgminMul<G, P>,
+    {
+        let zeros = param.zero_like();
+        let sig_param = P::max(&param.sub(&F::min_positive_value()).signum(), &zeros).add(&P::min(
+            &param.add(&F::min_positive_value()).signum(),
+            &zeros,
+        ));
+        self.xi = Some(
+            sig_param.add(
+                &sig_param
+                    .mul(&sig_param)
+                    .sub(&float!(1.0))
+                    .mul(pseudo_gradient),
+            ),
+        );
+        self.l1_coeff = Some(l1_coeff);
+    }
+}
+
+impl<O, P, G, F> CostFunction for LineSearchProblem<O, P, G, F>
+where
+    O: CostFunction<Param = P, Output = F>,
+    P: ArgminMul<P, P> + ArgminMinMax + ArgminSignum + ArgminZeroLike,
+{
+    type Param = P;
+    type Output = F;
+
+    fn cost(&self, param: &Self::Param) -> Result<Self::Output, Error> {
+        if let Some(xi) = self.xi.as_ref() {
+            let zeros = param.zero_like();
+            let param = P::max(&param.mul(xi).signum(), &zeros).mul(param);
+            self.problem.cost(&param)
```

Comment:
@stefan-k This line should be as follows:
```suggestion
            self.problem.cost(&param) + self.l1_coeff.unwrap() * param.l1_norm()
```
Should I add the l1_norm() function to the `ArgminNorm` trait? Or should I create a new trait for the L1 norm?

Reply:
I suggest adding a dedicated trait `ArgminL1Norm`. 
But then `ArgminNorm` should be renamed to `ArgminL2Norm`. If you want, you can also make that change, but it is absolutely fine if you don't want to do it because it is quite some work. I can also make that change after merging this :)